torch/_dynamo/__init__.py

```diff
@@ -157,13 +161,24 @@ def mark_dynamic(t, index):
     before torch.compile.
 
     """
+    mark_dynamic_constrained(t, index, min=None, max=None)
+
+
+@forbid_in_graph
+def mark_dynamic_constrained(
+    t, index, *, min: Optional[int] = None, max: Optional[int] = None
+):
     if isinstance(index, int):
         if not hasattr(t, "_dynamo_dynamic_indices"):
-            t._dynamo_dynamic_indices = set()
+            t._dynamo_dynamic_indices = defaultdict(MinMaxConstraint.NONE)
         # TODO(voz): Should we bounds check?
-        t._dynamo_dynamic_indices.add(index)
+        new_range = MinMaxConstraint(min=min, max=max)
+        curr_range = t._dynamo_dynamic_indices[index]
+        t._dynamo_dynamic_indices[index] = MinMaxConstraint.INTERSECT(
+            curr_range, new_range
```

Comment:
This actually needs to implement a disjoint check, since you can have disjoint ranges. It needs to:

1) check for overlap
2) If overlap, expand, if no overlap, new record

Reply:
Actually, no lol